### PR TITLE
🌪️ Support caching errors / exceptions

### PR DESCRIPTION
**Pull request by ronjouch**
This PR adds support for errors/exceptions caching.

- The default behavior is unchanged, and 100% backwards compatible.
- This PR only _adds_ an optional callback parameter, enabling callers to specify whether/when errors should be cached.

For use in our internal project where we want that.

---

**Note to reviewers**: remind this repo and PR are public, no talking about internal stuff in the review.

### CODE REVIEW

**Comment by ronjouch** [comment]
Louis/Eryk: actually, not ready yet. Will re-add you when it is

File: src/lib/CacheInstance.ts [hunk 1]
```diff
@@ -143,17 +149,35 @@ export abstract class CacheInstance extends EventEmitter {
...
+      }
+
+      // cache! results: always, errors: only if satisfying user assertion
+      let errorToCache: Error | undefined;
+      if (error && shouldCacheError(error)) {
+        errorToCache = error;
+      }
+      if ((errorToCache || result) !== undefined) {
+        await this.setValue(key, errorToCache || result, ttl);
+      }
```
Comment (erykwarren) [inline src/lib/CacheInstance.ts:177]: I think it is simpler with two separate calls:

```suggestion
      if (error && shouldCacheError(error)) {
        await this.setValue(key, errorToCache, ttl);
      } else if (result !== undefined) {
        await this.setValue(key, result, ttl);
      }
```

File: src/lib/RedisCache.ts [hunk 2]
```diff
@@ -130,6 +131,13 @@ export class RedisCache extends CacheInstance {
       return RedisCache.FALSE_VALUE;
     }
 
+    if (value instanceof Error) {
+      return RedisCache.ERROR_PREFIX + JSON.stringify({
+        ...value, // serialize potential Error metadata set as object properties
+        message: value.message,
```
Comment (erykwarren) [inline src/lib/RedisCache.ts:137]: That should be included in the previous line expansion, no?

File: src/lib/RedisCache.ts [hunk 3]
```diff
@@ -169,6 +177,16 @@ export class RedisCache extends CacheInstance {
...
     }
 
+    if (value.startsWith(RedisCache.ERROR_PREFIX)) {
+      const deserializedError = JSON.parse(value.substring(RedisCache.ERROR_PREFIX.length));
+      const error = new Error(deserializedError.message);
+      // restore potential Error metadata set as object properties
+      for (const [k, v] of Object.entries(deserializedError)) {
+        error[k] = v;
+      }
+      return error;
```
Comment (erykwarren) [inline src/lib/RedisCache.ts:187]: I believe you can simply do:
```suggestion
      return Object.assign(new Error(), deserializedError);
```

File: src/lib/RedisCache.ts (hunk 2 above)
Comment (ronjouch) [reply src/lib/RedisCache.ts:137]: > That should be included in the previous line expansion, no?

@erykwarren nope, an error's `message` is not a regular enumerable property.

```
Welcome to Node.js v14.2.0.
Type ".help" for more information.
> const e = new Error('hello Eryk')
undefined
> e.message
'hello Eryk'
> {...e}
{}
> e.author = 'Ronan'
'Ronan'
> e
Error: hello Eryk
    at repl:1:11
    at Script.runInThisContext (vm.js:131:20)
    at REPLServer.defaultEval (repl.js:436:29)
    at bound (domain.js:429:14)
    at REPLServer.runBound [as eval] (domain.js:442:12)
    at REPLServer.onLine (repl.js:763:10)
    at REPLServer.emit (events.js:327:22)
    at REPLServer.EventEmitter.emit (domain.js:485:12)
    at REPLServer.Interface._onLine (readline.js:337:10)
    at REPLServer.Interface._line (readline.js:666:8) {
  author: 'Ronan'
}
> {...e}
{ author: 'Ronan' }
> e.hasOwnProperty('author')
true
> e.hasOwnProperty('message')
true
```

I find the last two evaluations weird, I would have expected `false` for `e.hasOwnProperty('message')`. :shrug: , en tout cas, yes I need the explicit assignment for `message`.

File: src/lib/CacheInstance.ts (hunk 1 above)
Comment (ronjouch) [reply src/lib/CacheInstance.ts:177]: Hmmm I liked to express that caching errors vs. results took visibly the same path. But okay the "or" stuff is a little bit weird. 👍

**Pull request merged**